### PR TITLE
Announces winner or if there is a tie

### PR DESCRIPTION
**Pull request by mercebc**

### CODE REVIEW

File: lib/UI/UI.rb [hunk 1]
```diff
@@ -48,6 +48,17 @@ def show_error_message message
     @out.print message
   end
 
+  def announce message
+    @out.print message
+  end
+
+  def announce_results(board, players)
+    announce("It's a tie") if board.is_full?
```
Comment (cmvandrevala) [inline lib/UI/UI.rb]: I think the Ruby convention is to use `full?` rather than `is_full?`

File: lib/UI/UI.rb [hunk 2]
```diff
@@ -48,6 +48,17 @@ def show_error_message message
...
+    for player in players do
+      announce(player.symbol + " has won!") if board.is_a_winner?(player)
```
Comment (cmvandrevala) [inline lib/UI/UI.rb:59]: Same here. I think this would be `board.winner?`

File: lib/core/board.rb [hunk 3]
```diff
@@ -5,27 +5,82 @@ class Board
   attr_reader :capacity
   attr_reader :size
```
Comment (cmvandrevala) [inline lib/core/board.rb]: You can combine these two readers onto one line: `attr_reader :capacity, :size`

File: lib/UI/UI.rb [hunk 4]
```diff
@@ -48,6 +48,17 @@ def show_error_message message
     @out.print message
   end
 
+  def announce message
+    @out.print message
+  end
+
+  def announce_results(board, players)
+    announce("It's a tie") if board.is_full?
+    for player in players do
```
Comment (ssunday) [inline lib/UI/UI.rb]: players.each is more idiomatic

File: lib/core/board.rb [hunk 5]
```diff
@@ -5,27 +5,82 @@ class Board
...
   end
 
-  def invalid_cell position
+  def invalid_cell(position)
     !@grid[position].is_empty?
   end
-end
 
+  def has_a_winner?(players)
+    for player in players do
```
Comment (ssunday) [inline lib/core/board.rb]: players.any? { |player| is_a_winner?(player) }

File: lib/core/board.rb [hunk 6]
```diff
@@ -5,27 +5,82 @@ class Board
...
+      return true if is_a_winner?(player)
+    end
+    false
```
Comment (cmvandrevala) [inline lib/core/board.rb]: This method can probably be simplified by the use of `select`: https://ruby-doc.org/core-2.2.0/Array.html#method-i-select

File: lib/core/board.rb [hunk 7]
```diff
@@ -5,27 +5,82 @@ class Board
...
+  end
+
+  def is_a_winner?(player)
+    lines = get_all_lines
+    for line in lines do
```
Comment (ssunday) [inline lib/core/board.rb]: using each/map/any/all is cleaner than using a for loop and returning true/false

File: lib/core/board.rb [hunk 8]
```diff
@@ -5,27 +5,82 @@ class Board
...
+      right_diagonal << rows[num][idx]
+    end
+    right_diagonal
+  end
+
+  def get_all_lines
+    lines = Array.new
+    columns = get_columns
+    rows = get_rows
+    for line in columns do
```
Comment (ssunday) [inline lib/core/board.rb]: Mapping could be used here

File: lib/core/board.rb [hunk 9]
```diff
@@ -5,27 +5,82 @@ class Board
...
+    end
+    false
+  end
+
+  def is_a_winner?(player)
+    lines = get_all_lines
+    for line in lines do
+      return true if has_a_winning_line(line, player)
+    end
+    false
```
Comment (cmvandrevala) [inline lib/core/board.rb]: You might be able to use `select` here as well

File: lib/core/board.rb [hunk 10]
```diff
@@ -5,27 +5,82 @@ class Board
...
+    rows = get_rows
+    for line in columns do
+      lines << line
+    end
+    for line in rows do
+      lines << line
+    end
+    lines << get_left_diagonal(columns)
+    lines << get_right_diagonal(rows)
+    lines
```
Comment (cmvandrevala) [inline lib/core/board.rb]: You might be able to do something like this here:

```
get_columns.push(get_rows).push(get_left_diagonal).push(get_right_diagonal)
```

File: lib/core/cell.rb [hunk 11]
```diff
@@ -9,4 +9,8 @@ def initialize
   def is_empty?
     @content == EMPTY
   end
+
+  def belongs_to(player)
+    @content == player.symbol
```
Comment (cmvandrevala) [inline lib/core/cell.rb:14]: This should probably be `belongs_to?` as per Ruby conventions

File: spec/UI/UI_spec.rb [hunk 12]
```diff
@@ -14,19 +20,40 @@
...
-    user_input = StringIO.new
-    ui = UI.new(output, user_input)
-    board = Board.new(3)
+    ui = UI.new(output, StringIO.new)
     board.mark(3,"X")
     ui.show_grid(board)
     expect(output.string).to eq("   |   |   \n---+---+---\n X |   |   \n---+---+---\n   |   |   \n")
   end
 
+  it 'announce a tie' do
```
Comment (cmvandrevala) [inline spec/UI/UI_spec.rb]: Try to make the description match the "it" matcher. For example:

```
it "announces a tie" do ...
```

File: spec/core/board_spec.rb [hunk 13]
```diff
@@ -6,6 +6,12 @@
 
   let(:board) { Board.new(3) }
 
+  def mark_board(marks)
+    for position in marks
+      board.mark(position, "X")
+    end
+  end
+
```
Comment (cmvandrevala) [inline spec/core/board_spec.rb:14]: This helper method is duplicated between two files. Can you refactor your tests to pull it out into its own file?

**Review by cmvandrevala — COMMENTED** [review]
Overall this is a pretty good PR. Most of my comments focused on formatting and Ruby conventions. Nice work!

File: spec/core/flow_spec.rb [hunk 14]
```diff
@@ -5,35 +5,38 @@
...
-    ui = UI.new
-    flow = Flow.new(ui, board)
     flow.swap_players
     expect(flow.current_player.symbol).to eq("O")
     expect(flow.opponent.symbol).to eq("X")
   end
 
+  it 'game is over' do
+    mark_board([0, 1, 2])
+    expect(flow.game_over).to be true
```
Comment (ssunday) [inline spec/core/flow_spec.rb]: Be consistent about the parenthesis usage for the be/eq (also with be vs. eq)

File: spec/core/flow_spec.rb [hunk 15]
```diff
@@ -5,35 +5,38 @@
...
-    user_input = StringIO.new("1\n1\n2\n3\n4\n5\n6\n7\n8\n9")
-    ui = UI.new(output, user_input)
-    flow = Flow.new(ui, board)
-    flow.start
-    expect(output.string).to include("Please choose a different position where the cell is empty\n")
-  end
+    let(:flow) { Flow.new(ui, board) }
+
+    def mark_board(marks)
+      for position in marks
```
Comment (ssunday) [inline spec/core/flow_spec.rb]: Each is more idiomatic, etc, etc.

File: spec/core/flow_spec.rb (hunk 14 above)
Comment (mercebc) [reply spec/core/flow_spec.rb]: is it the same? when do you use be vs eq? 
I will put the parenthesis now, thanks!!

File: spec/core/flow_spec.rb (hunk 15 above)
Comment (mercebc) [reply spec/core/flow_spec.rb]: sure!! thanks!

File: spec/core/flow_spec.rb (hunk 14 above)
Comment (ssunday) [reply spec/core/flow_spec.rb]: Personally, never used `be`, always eq

**Pull request merged**